packages/pkcs12/src/index.ts

```diff
@@ -0,0 +1,106 @@
+import crypto, { HexBase64Latin1Encoding, Utf8AsciiLatin1Encoding } from 'crypto';
+import forge from 'node-forge';
+
+/**
+ * returns serial number as an uppercased hexidecimal string
+ */
+export function getFormattedSerialNumber(certificate: forge.pki.Certificate): string | null {
+  const { serialNumber } = certificate;
+  return serialNumber ? serialNumber.replace(/^0+/, '').toUpperCase() : null;
+}
+
+/**
+ * Extracts a certificate from PKCS#12
+ * This is assumed to be a conventional PKCS#12 where there is exactly one certificate and one key
+ */
+export function getX509Certificate(p12: forge.pkcs12.Pkcs12Pfx): forge.pki.Certificate {
+  const certBagType = forge.pki.oids.certBag;
+  const bags = p12.getBags({ bagType: certBagType })[certBagType];
+  if (!bags || bags.length === 0) {
+    throw new Error(`PKCS12: No certificates found`);
+  }
+  const certificate = bags[0].cert;
+  if (!certificate) {
+    throw new Error('PKCS12: bag is not a certificate');
+  }
+  return certificate;
+}
+
+/**
+ * Extracts a certificate from PKCS#12
+ * This is assumed to be a PKCS#12 in Keystore format where the friendlyName (alias) contains a PrivateKeyEntry
+ * A PrivateKeyEntry contains exactly one certificate and one key
+ * https://docs.oracle.com/javase/7/docs/api/java/security/KeyStore.PrivateKeyEntry.html
+ */
+export function getX509CertificateByFriendlyName(
+  p12: forge.pkcs12.Pkcs12Pfx,
+  friendlyName: string
+): forge.pki.Certificate {
+  const certBagType = forge.pki.oids.certBag;
+  const bags = p12.getBags({ friendlyName, bagType: certBagType }).friendlyName;
+  if (!bags || bags.length === 0) {
+    throw new Error(`PKCS12: No certificates found under friendlyName: ${friendlyName}`);
+  }
+  const certificate = bags[0].cert;
+  if (!certificate) {
+    throw new Error('PKCS12: bag is not a certificate');
+  }
+  return certificate;
+}
+
+export function getPKCS12(
+  p12BufferOrBase64String: Buffer | string,
+  maybePassword: string | null
+): forge.pkcs12.Pkcs12Pfx {
+  const base64EncodedP12 = Buffer.isBuffer(p12BufferOrBase64String)
+    ? p12BufferOrBase64String.toString('base64')
+    : p12BufferOrBase64String;
+  const password = String(maybePassword || '');
+  const p12Der = forge.util.decode64(base64EncodedP12);
```

Comment:
what is the format of p12Der? just want to make sure we're not doing extra work converting a buffer to base64 and then converting it back to some buffer-like data type.

Reply:
p12Der is a binary encoded string. The way we are getting the p12 file is how `node-forge` recommends in their [docs](https://github.com/digitalbazaar/forge#pkcs12). The methods they expose are basically to get the base64 p12 -> decode it into a binary encoded string (p12Der) -> convert it to an asn1 object -> convert it to a pkcs12 object.